overlord/snapstate/handlers_components.go

```diff
@@ -227,3 +240,169 @@ func (m *SnapManager) undoMountComponent(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// snapSt is a copy of the current state
+	compSetup, snapsup, snapSt, err := compSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	csi := compSetup.CompSideInfo
+
+	// Grab information for current snap revision
+	// TODO will this still be correct when jointly installing snap +
+	// components? Link for the snap should happen earlier. But we do not
+	// want to reboot until all components are linked to the initramfs has
+	// the information to moun them already.
+	snapInfo, err := snapSt.CurrentInfo()
+	// the owner snap is expected to be installed (ErrNoCurrent not allowed)
+	if err != nil {
+		return err
+	}
+	snapIdx := snapSt.LastIndex(snapInfo.Revision)
+	revSt := snapSt.Sequence.Revisions[snapIdx]
+
+	// set information for undoLinkComponent in the task
+	t.Set("linked-component", csi)
+	// Append new component to components of the current snap
+	revSt.Components = append(revSt.Components, csi)
+
+	// Store last component refresh time
+	if snapSt.LastCompRefreshTime == nil {
+		snapSt.LastCompRefreshTime = make(map[string]time.Time)
+	}
+	snapSt.LastCompRefreshTime[csi.Component.ComponentName] = timeNow()
+
+	// Finally, write the state
+	Set(st, snapsup.InstanceName(), snapSt)
+
+	return nil
+}
+
+func (m *SnapManager) undoLinkComponent(t *state.Task, _ *tomb.Tomb) error {
+	// invariant: component is installed
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// snapSt is a copy of the current state
+	_, snapsup, snapSt, err := compSetupAndState(t)
+	if err != nil {
+		return err
+	}
+
+	// Expected to be installed
+	snapInfo, err := snapSt.CurrentInfo()
+	if err != nil {
+		return err
+	}
+	snapIdx := snapSt.LastIndex(snapInfo.Revision)
+	revSt := snapSt.Sequence.Revisions[snapIdx]
+
+	var linkedComp *snap.ComponentSideInfo
+	err = t.Get("linked-component", &linkedComp)
+	if err != nil {
+		return err
+	}
+
+	// Restore old state
+	// relinking of the old component is done in the undo of unlink-current-snap
+	restoredComps := []*snap.ComponentSideInfo{}
+	for _, csi := range revSt.Components {
+		if csi.Equal(linkedComp) {
+			continue
+		}
+		restoredComps = append(restoredComps, csi)
+	}
+	revSt.Components = restoredComps
+
+	// Finally, write the state
+	Set(st, snapsup.InstanceName(), snapSt)
+
+	return nil
+}
+
+func (m *SnapManager) doUnlinkCurrentComponent(t *state.Task, _ *tomb.Tomb) (err error) {
+	// invariant: current snap has a revision for this component installed
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// snapSt is a copy of the current state
+	compSetup, snapsup, snapSt, err := compSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	cref := compSetup.CompSideInfo.Component
+
+	// Expected to be installed
+	snapInfo, err := snapSt.CurrentInfo()
+	if err != nil {
+		return err
+	}
+
+	// Remove current component for the current snap
+	snapIdx := snapSt.LastIndex(snapInfo.Revision)
+	revSt := snapSt.Sequence.Revisions[snapIdx]
```

Comment:
as discussed previously some of the code here seem that it could be methods on SnapSequence